add API Reference(draft) #2

takayukifujisawa opened the pull request:
12/7中くらいまでに概ねfixしたいです

kawanamiyuu commented on README.md:
````diff
@@ -1 +1,176 @@
+]
+```
+
+### ログイン [GET /login/v/{verion}/name/{name}/email/{email}/]
````
これはログイン処理をするためのエンドポイントと思いますが、とすると、
`POST /login/v/{verion}/`ではないでしょうか。`name`と`email`はリクエストボディでパラメータとしてサーバに送信するのがよいと思います。

`/me/v/{v}/token/{token}`はリソース（この場合はユーザ）を一意に特定しますが、`/login`はログインというアクションなので、リソースではないと考えます。

takayukifujisawa replied:
上記の通り対応予定。

kawanamiyuu commented on README.md:
```diff
@@ -1 +1,176 @@
+    	"email": "foo@example.com",
+    }
+    "_links": {
+        "self": { "href": "/login/v/0.1/name/{name}/email/{email}/" },
```
このレスポンスオブジェクトが「ユーザ」とすると、この場合の`self`はログインしたユーザ自身を表す`/me/v/{v}/token/{token}`になります。

takayukifujisawa replied:
上記の通り対応予定。

kawanamiyuu commented on README.md:
```diff
@@ -1 +1,176 @@
+[
+  {
+    "version": "0.1",
+    "_links": {
```
`_links`にはそのリソースから関連のあるリソースへの`リンク(URI)`を記述します。
※エンドポイントの一覧ではないので注意が必要です。
ですので、リソースでなく、アクション(操作)を表す`login`,`logout`,`join`や、`/`から直接関連のない`me`も`_links`には記述しないのが自然と考えます。
と考えると、`/`は何を表すURIなのか（`/`というエンドポイントは必要か）というところを明確にする必要がありそうです。

エンドポイントを設計する上で、それが`リソース`なのか、`アクション(操作)`なのかを意識する必要がありそうです。

takayukifujisawa replied:
一旦`/`は消しましょうか。`login`にしろ`calendar`にしろ、クライアント側に固定のURLを保持しておかないと
いけないのがHAL的にどうなんだろう？ってところから`/`に主要なリンクをくっつけたんですが、
根拠もふわっとしてるので一旦消しちゃいますね。

kawanamiyuu commented on README.md:
```diff
@@ -1 +1,176 @@
+        "logout" : { "href": "/logout" },
+        "me" : { "href": "/me/v/0.1/token/{token}" },
+    },
+    "_embeded": {
```
`_embeded`には、そのリソースに関連する別のリソースオブジェクトを埋め込みます。
`days`はまさにカレンダーオブジェクトそのものですので、`_embeded`の外に記述するのが自然です。

takayukifujisawa replied:
上記の通り対応予定。

kawanamiyuu commented on README.md:
```diff
@@ -1 +1,176 @@
+    "_embeded": {
+    	// 一ヶ月分のカレンダーと各イベントの状態
+    	"days" {
+    		"1":{
```
好みの問題もありますが、JSONオブジェクトのkeyが数字なのはちょっと気になります。
`days`がdayオブジェクトの配列で、そのdayオブジェクトが`dayOfMonth: 1`のような感じで日にちを保持しているというのはどうでしょうか。

``` js
"days": [
  {
    "dayOfMonth": 1,
        :
  },
  {
    "dayOfMonth": 2,
        :
  }
]
```

takayukifujisawa replied:
こっちのほうがスマートですね。上記の通り対応予定。

kawanamiyuu commented on README.md:
```diff
@@ -1 +1,176 @@
+    		"1":{
+    			"doy": "SUN",
+    			"events" {
+    				"lunch" : {
```
(要追加) 自分が参加しているかどうかという情報

takayukifujisawa replied:
上記の通り対応予定。

takayukifujisawa commented:
あと、肝心のグループIDを忘れていたので、それも追記しますね。

takayukifujisawa committed:
fix API Reference following PR #2

 common - レスポンスをキャメルケースに統一
 		  _linksにmethodとparametersを追加(post用)
 root - 削除
 login - GETからPOSTへ。
         parameterにグループIDを追加。
         selfをmeに。(ルー大柴か)
 calendar - レスポンスに「既に参加済みか」フラグ追加
 join - GETからPOSTへ。

takayukifujisawa committed:
fix API Reference following PR #2

common - バージョンのパラメータを`v`で受けるのにレスポンスが`version`で微妙だったので修正
                 _links精査。必要最小限に。

takayukifujisawa commented:
再チェックお願いします！

kawanamiyuu commented:
カレンダーの綴りはcalend `a` rが正ですね

takayukifujisawa commented:
これは恥ずかしい。なおしますー。

takayukifujisawa committed:
modify Api reference #2

takayukifujisawa committed:
typo #2

takayukifujisawa committed:
Merge branch 'master' into documentation

takayukifujisawa committed:
change entry point for joining #2

kawanamiyuu commented:
![LGTM](https://40.media.tumblr.com/39dc44d16604337beb1c3f9cc6817998/tumblr_nw20cwm5p01uyzky1o1_500.jpg)
